src/mpid/ch4/generic/mpidig_msg.h

```diff
@@ -0,0 +1,558 @@
+                                                            rreq->status.MPI_TAG, data_sz,
+                                                            in_data_sz);
+            }
+
+            MPIR_STATUS_SET_COUNT(rreq->status, done);
+        }
+    }
+}
+
+static inline int MPIDIG_match_msg(int rank, int tag, MPIR_Context_id_t context_id,
+                                   int error_bits, size_t data_sz, int is_local,
+                                   int *got_match, MPIR_Request ** req)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIR_Request *rreq = NULL;
+    MPIR_Comm *root_comm;
+
+    root_comm = MPIDIG_context_id_to_comm(context_id);
+    if (root_comm) {
+        rreq = MPIDIGI_match_posted(rank, tag, context_id, root_comm);
+    }
+
+    if (rreq == NULL) {
+        mpi_errno = MPIDIGI_enqueue_unexp(rank, tag, context_id, root_comm, data_sz, is_local, &rreq);
+        MPIR_ERR_CHECK(mpi_errno);
+        *got_match = 0;
+    } else {
+        /* rreq != NULL <=> root_comm != NULL */
+        MPIR_Assert(root_comm);
+        /* Decrement the refcnt when popping a request out from posted_list */
+        MPIR_Comm_release(root_comm);
+        MPIDIG_REQUEST(rreq, rank) = rank;
+        MPIDIG_REQUEST(rreq, tag) = tag;
+        MPIDIG_REQUEST(rreq, context_id) = context_id;
+        *got_match = 1;
+    }
+
+    *req = rreq;
+
+  fn_exit:
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+static inline int MPIDIG_handle_send(int rank, int tag, MPIR_Context_id_t context_id,
+                                     int error_bits, void **data, size_t * p_data_sz,
+                                     int is_local, int *is_contig,
+                                     MPIDIG_am_target_cmpl_cb * target_cmpl_cb,
+                                     MPIR_Request ** req)
+{
+    int mpi_errno = MPI_SUCCESS;
+
+    void *in_data = *data;
+    size_t in_data_sz = p_data_sz ? *p_data_sz : 0;
+
+    int got_match;
+    MPIR_Request * rreq;
+    mpi_errno = MPIDIG_match_msg(rank, tag, context_id, error_bits, in_data_sz, is_local,
+                                 &got_match, &rreq);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    rreq->status.MPI_ERROR = error_bits;
+    MPIDIG_REQUEST(rreq, req->status) |= MPIDIG_REQ_IN_PROGRESS;
+
+    mpi_errno = do_send_target(data, p_data_sz, is_contig, target_cmpl_cb, rreq);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    /* TODO:: 
+    MPIDIG_recv_copy(in_data, in_data_sz, *data, *p_data_sz, is_contig, rreq);
+    recv_target_cmpl_cb(rreq);
+     */
+
+#ifndef MPIDI_CH4_DIRECT_NETMOD
+    if (unlikely(MPIDI_REQUEST_ANYSOURCE_PARTNER(rreq))) {
+        anysrc_free_partner(rreq);
+    }
+#endif /* MPIDI_CH4_DIRECT_NETMOD */
+
+    *req = rreq;
+  fn_exit:
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+int MPIDIG_reply_ssend(MPIR_Request * rreq);
+/* This function is called when a receive has completed on the receiver side. The input is the
```

Comment:
ssend ack can be sent once the message has matched. It does not need to wait for recv completion.

Reply:
In ch3, we only differentiated eager ssends. Rendezvous ssends are defined as synchronous assuming the CTS message is sent after the message is matched.

Reply:
Are you saying that rendezvous can serve as ssend?

It appears that we don't do rendezvous for ssend in ch4. I'm thinking that we need a separate ssend rendezvous protocol rather than nesting protocols to allow more precise implementations.

Reply:
Yes, rendezvous as implemented today is synchronous. No extra ack is needed for rndv protocol messages. Really what I am pointing out is that waiting for the entire message to be received is overkill with regard to MPI requirements. MPI only mandates the message is matched by the receiver, which comes in the form of an ACK for eager messages, or a CTS for rndv.